includes/class-linkbacks-mf2-handler.php

```diff
@@ -106,137 +113,279 @@ public static function get_rel_mapper() {
+		if ( isset( $properties[ $key ] ) && isset( $properties[ $key ][0] ) ) {
+			if ( is_array( $properties[ $key ] ) ) {
+				$properties[ $key ] = array_unique( $properties[ $key ] );
 			}
+			if ( 1 === count( $properties[ $key ] ) ) {
+				return $properties[ $key ][0];
+			}
+			return $properties[ $key ];
 		}
+		return null;
+	}
 
-		// set canonical url (u-url)
-		if ( self::check_mf_attr( 'url', $properties ) ) {
-			$commentdata['comment_meta']['semantic_linkbacks_canonical'] = esc_url_raw( $properties['url'][0] );
-		} else {
-			$commentdata['comment_meta']['semantic_linkbacks_canonical'] = esc_url_raw( $source );
+	/**
+	 * Is string a URL.
+		 *
+	 * @param array $string
+	 * @return bool
+	 */
+	public static function is_url( $string ) {
+		if ( ! is_string( $string ) ) {
+			return false;
 		}
+		return preg_match( '/^https?:\/\/.+\..+$/', $string );
+	}
 
-		// check rsvp property
-		if ( self::check_mf_attr( 'rsvp', $properties ) ) {
-			$commentdata['comment_meta']['semantic_linkbacks_type'] = wp_slash( 'rsvp:' . $properties['rsvp'][0] );
+	// Accepted h types
+	public static function is_h( $string ) {
+		return in_array( $string, array( 'h-cite', 'h-entry', 'h-feed', 'h-product', 'h-event', 'h-review', 'h-recipe' ) );
+	}
+
+	public static function flatten_microformats( $item ) {
+		$flat = array();
+		if ( 1 === count( $item ) ) {
+			$item = $item[0];
+		}
+		if ( array_key_exists( 'type', $item ) ) {
+			// If there are multiple types strip out everything but the standard one.
+			if ( 1 < count( $item['type'] ) ) {
+				$item['type'] = array_filter( $item['type'], array( 'Linkbacks_MF2_Handler', 'is_h' ) );
+			}
+			$flat['type'] = $item['type'][0];
+		}
+		if ( array_key_exists( 'properties', $item ) ) {
+			$properties = $item['properties'];
+			foreach ( $properties as $key => $value ) {
+				$flat[ $key ] = self::get_property( $key, $properties );
+				if ( 1 < count( $flat[ $key ] ) ) {
+					$flat[ $key ] = self::flatten_microformats( $flat[ $key ] );
+				}
+			}
 		} else {
-			// get post type
-			$commentdata['comment_meta']['semantic_linkbacks_type'] = wp_slash( self::get_entry_type( $commentdata['target'], $entry, $mf_array ) );
+			$flat = $item;
 		}
-
-		return $commentdata;
+		foreach ( $flat as $key => $value ) {
+			// Sanitize all URL properties
+			if ( self::is_url( $value ) ) {
+				$flat[ $key ] = esc_url_raw( $value );
+			}
+		}
+		return array_filter( $flat );
 	}
 
 	/**
-	 * get all h-entry items
+	 * helper to find the representative item
 	 *
-	 * @param array $mf_array the microformats array
-	 * @param array the h-entry array
+	 * @param array $mf_array the parsed microformats array
+	 * @param string $target the target url
 	 *
-	 * @return array
+	 * @return array representative item or false if none
 	 */
-	public static function get_entries( $mf_array ) {
-		$entries = array();
-
+	public static function get_representative_item( $mf_array, $target ) {
```

Comment:
this is called for getting the representative h-entry and h-card, so we shouldn't mix both. This might cause errors because it might return the author as h-entry and the other way around. The author stuff should be part of `get_representative_author`.